libvast/vast/detail/generator.hpp

```diff
@@ -0,0 +1,270 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+// cppcoro, adapted for libvast.
+//
+// Original Author: Lewis Baker
+// Original License: MIT
+// Original Revision: a87e97fe5b6091ca9f6de4637736b8e0d8b109cf
+// Includes code from https://github.com/lewissbaker/cppcoro
```

Comment:
Have you considered using the variant proposed in p2168r1 instead? https://godbolt.org/z/T58h1W 

It supports recursion on `generator<T>` which could simplify some of the implemented functions like `leaves()` considerably.

Otoh it seems that proposal is being held up in LEWG, bike scheds are being painted...

Reply:
See comparison with the recursive version here: https://github.com/lewissbaker/cppcoro#recursive_generatort

To me it is not clear which of the two will end up being in the standup. The cppcoro one is battle-tested, so right now I'd prefer using that. The smaller the coroutine stack, the bigger the likelihood of the compiler being able to elide it.

Reply:
For the record, the implementation from p2168 is different than the one in cppcoro. The latter uses a dedicated `recursive_generator` while the format supports recursion with the regular `generator`.